File: frontend/js/views/crypto_liquidation.js
// Crypto perpetual liquidation price (isolated margin), via /calc/crypto-liquidation.
import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');
export async function renderCryptoLiquidation(mount, _s) {
    const tok = currentViewToken(); if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.cliq.h1.title">// PERP LIQUIDATION PRICE</span></h1>
        <p class="muted small" data-i18n="view.cliq.hint.intro">The mark price at which an isolated-margin linear (USDT) perpetual position is force-liquidated. Long liquidation = entry × (1 − 1/leverage + maintenance margin rate); short flips the signs. Cross-margin and inverse contracts differ. Not financial advice.</p>
        <div class="lpv-split"><div class="chart-panel"><h2 data-i18n="view.cliq.h2.inputs">Position</h2>
        <form id="cliq-form" class="inline-form">
            <label><span data-i18n="view.cliq.label.side">Side</span><select name="side"><option value="long" data-i18n="view.cliq.opt.long">Long</option><option value="short" data-i18n="view.cliq.opt.short">Short</option></select></label>
            <label><span data-i18n="view.cliq.label.entry">Entry price ($)</span><input type="number" step="0.01" min="0" name="entry_price_usd" value="30000" required></label>
            <label><span data-i18n="view.cliq.label.lev">Leverage (×)</span><input type="number" step="1" min="1" name="leverage" value="10" required></label>
            <label><span data-i18n="view.cliq.label.mmr">Maint. margin rate (decimal)</span><input type="number" step="0.001" min="0" name="maintenance_margin_rate" value="0.005"></label>
            <label><span data-i18n="view.cliq.label.size">Position size (coins)</span><input type="number" step="0.1" min="0" name="position_size" value="1"></label>
        </form></div><div id="cliq-result" class="chart-panel lpv-preview"></div></div>`;
    applyUiI18n(mount);
    const form = mount.querySelector('#cliq-form'); const n = (k) => Number(form.querySelector(`[name="${k}"]`).value) || 0;
    const gen = async () => {
        const body = { side: form.querySelector('[name="side"]').value, entry_price_usd: n('entry_price_usd'), leverage: n('leverage'), maintenance_margin_rate: n('maintenance_margin_rate'), position_size: n('position_size') };
        try { const d = await api.calcCryptoLiquidation(body); if (!viewIsCurrent(tok)) return; res(mount, d); }
        catch (e) { showToast(e.message || t('view.cliq.toast.error'), { level: 'error' }); }
    };
    const live = debounce(gen, 250); form.addEventListener('input', () => live()); form.addEventListener('submit', (e) => { e.preventDefault(); gen(); }); gen();
}
function res(mount, d) {
    const el = mount.querySelector('#cliq-result');
    if (!d.valid) { el.innerHTML = `<p class="muted" data-i18n="view.cliq.invalid">Invalid inputs (leverage and entry must be positive).</p>`; applyUiI18n(el); return; }
    el.innerHTML = `<div class="lpv-bar"><div class="cards">
        <div class="card neg"><div class="label" data-i18n="view.cliq.card.liq">Liquidation price</div><div class="value">${money(d.liquidation_price_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.cliq.card.dist">Distance to liq</div><div class="value">${pct(d.distance_pct)}</div></div>
        <div class="card"><div class="label" data-i18n="view.cliq.card.bank">Bankruptcy price</div><div class="value">${money(d.bankruptcy_price_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.cliq.card.im">Initial margin</div><div class="value">${money(d.initial_margin_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.cliq.card.mm">Maint. margin</div><div class="value">${money(d.maintenance_margin_usd)}</div></div>
    </div></div>`;
    applyUiI18n(el);
}

File: crates/traderview-core/src/crypto_liquidation.rs
//! Crypto perpetual liquidation price (isolated margin, linear/USDT-settled) — the
//! mark price at which an isolated-margin perpetual position is force-liquidated.
//! For a long, liquidation is `entry × (1 − 1/leverage + maintenance_margin_rate)`;
//! for a short, `entry × (1 + 1/leverage − maintenance_margin_rate)`. It also
//! reports the distance to liquidation, notional, and the initial/maintenance
//! margin. Pure compute, not exchange-specific (cross-margin and inverse contracts
//! differ). Not financial advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct LiquidationInput {
    /// "long" or "short".
    pub side: String,
    pub entry_price_usd: f64,
    /// Leverage, e.g. 10 for 10×.
    pub leverage: f64,
    /// Maintenance margin rate, decimal (e.g. 0.005 = 0.5%).
    #[serde(default = "default_mmr")]
    pub maintenance_margin_rate: f64,
    /// Position size in coins/contracts (for margin amounts).
    #[serde(default)]
    pub position_size: f64,
}

fn default_mmr() -> f64 {
    0.005
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct LiquidationReport {
    pub is_long: bool,
    pub liquidation_price_usd: f64,
    /// Bankruptcy price (equity = 0, ignoring maintenance margin).
    pub bankruptcy_price_usd: f64,
    /// |liq − entry| ÷ entry, percent.
    pub distance_pct: f64,
    pub notional_usd: f64,
    pub initial_margin_usd: f64,
    pub maintenance_margin_usd: f64,
    pub valid: bool,
}

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

pub fn generate(i: &LiquidationInput) -> LiquidationReport {
    if i.leverage <= 0.0 || i.entry_price_usd <= 0.0 {
        return LiquidationReport::default();
    }
    let is_long = !i.side.trim().eq_ignore_ascii_case("short");
    let imr = 1.0 / i.leverage;
    let mmr = i.maintenance_margin_rate;
    let liq = if is_long {
        i.entry_price_usd * (1.0 - imr + mmr)
    } else {
        i.entry_price_usd * (1.0 + imr - mmr)
    };
    let bankruptcy = if is_long {
        i.entry_price_usd * (1.0 - imr)
    } else {
        i.entry_price_usd * (1.0 + imr)
    };
    let notional = i.entry_price_usd * i.position_size;
    LiquidationReport {
        is_long,
        liquidation_price_usd: round2(liq),
        bankruptcy_price_usd: round2(bankruptcy.max(0.0)),
        distance_pct: round2((liq - i.entry_price_usd).abs() / i.entry_price_usd * 100.0),
        notional_usd: round2(notional),
        initial_margin_usd: round2(notional * imr),
        maintenance_margin_usd: round2(notional * mmr),
        valid: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> LiquidationInput {
        LiquidationInput {
            side: "long".into(),
            entry_price_usd: 30_000.0,
            leverage: 10.0,
            maintenance_margin_rate: 0.005,
            position_size: 1.0,
        }
    }

    #[test]
    fn long_liquidation() {
        let d = generate(&base());
        assert!(d.is_long);
        assert!(close(d.liquidation_price_usd, 27_150.0));
        assert!(close(d.distance_pct, 9.5));
        assert!(close(d.bankruptcy_price_usd, 27_000.0));
        assert!(close(d.initial_margin_usd, 3_000.0));
        assert!(close(d.maintenance_margin_usd, 150.0));
    }

    #[test]
    fn short_liquidation() {
        let d = generate(&LiquidationInput { side: "short".into(), ..base() });
        assert!(!d.is_long);
        assert!(close(d.liquidation_price_usd, 32_850.0));
        assert!(close(d.distance_pct, 9.5));
    }

    #[test]
    fn higher_leverage_closer_liquidation() {
        let d = generate(&LiquidationInput { leverage: 50.0, ..base() });
        // 30000 × (1 − 0.02 + 0.005) = 29550.
        assert!(close(d.liquidation_price_usd, 29_550.0));
        assert!(d.distance_pct < 9.5);
    }

    #[test]
    fn invalid_leverage() {
        let d = generate(&LiquidationInput { leverage: 0.0, ..base() });
        assert!(!d.valid);
    }
}
